File: ploy/experiment/src/die/DieSides.py
from __future__ import annotations
from enum import Enum
import random


class DieSides(Enum):
    NONE = 0
    D4 = 4
    D6 = 6
    D8 = 8
    D10 = 10
    D12 = 12
    D20 = 20

    def roll(self) -> int:
        if self == DieSides.NONE:
            return 0
        return random.randint(1, self.value)

    @property
    def sides(self) -> int:
        return self.value

    def downgrade(self) -> DieSides:
        if self == DieSides.D20:
            return DieSides.D12
        if self == DieSides.D12:
            return DieSides.D10
        if self == DieSides.D10:
            return DieSides.D8
        if self == DieSides.D8:
            return DieSides.D6
        if self == DieSides.D6:
            return DieSides.D4
        return DieSides.NONE

    def average(self) -> float:
        if self.value == 0:
            return 0
        return (self.value + 1) / 2

File: leaf14/src/main/kotlin/dugsolutions/leaf/random/die/DieSides.kt
package dugsolutions.leaf.random.die

enum class DieSides(val value: Int) {
    D4(4),
    D6(6),
    D8(8),
    D10(10),
    D12(12),
    D20(20);

    companion object {
        fun from(sides: Int): DieSides {
            return when (sides) {
                4 -> D4
                6 -> D6
                8 -> D8
                10 -> D10
                12 -> D12
                20 -> D20
                else -> throw Exception("Unknown number of sides: $sides")
            }
        }

    }

    override fun toString(): String {
        return "D$value"
    }

    val downgrade: DieSides?
        get() {
            return when(value) {
                4 -> null
                6 -> D4
                8 -> D6
                10 -> D8
                12 -> D10
                20 -> D12
                else -> null
            }
        }

    val upgrade: DieSides?
        get() {
            return when(value) {
                4 -> D6
                6 -> D8
                8 -> D10
                10 -> D12
                12 -> D20
                else -> null
            }
        }

}
